src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs

```diff
@@ -795,39 +817,53 @@ private async ValueTask<int> ReadAsyncInternal<TIOAdapter>(TIOAdapter adapter, M
                         int readBytes = await adapter.ReadAsync(_internalBuffer.AsMemory(_internalBufferCount)).ConfigureAwait(false);
                         if (readBytes == 0)
                         {
-                            return 0;
+                            receivedEOF = true;
+                            break;
                         }
 
                         _internalBufferCount += readBytes;
                         if (_internalBufferCount < SecureChannel.ReadHeaderSize)
                         {
                             await FillBufferAsync(adapter, SecureChannel.ReadHeaderSize).ConfigureAwait(false);
                         }
-                    }
-                    Debug.Assert(_internalBufferCount >= SecureChannel.ReadHeaderSize);
 
-                    // Parse the frame header to determine the payload size (which includes the header size).
-                    int payloadBytes = GetFrameSize(_internalBuffer.AsSpan(_internalOffset));
-                    if (payloadBytes < 0)
-                    {
-                        throw new IOException(SR.net_frame_read_size);
-                    }
+                        if (payloadBytes == int.MaxValue)
+                        {
+                            // Parse the frame header to determine the payload size unless already done above.
+//                            payloadBytes = GetFrameSize(_internalBuffer.Span.Slice(_internalOffset));
+                            payloadBytes = GetFrameSize(_internalBuffer.AsSpan(_internalOffset));
+                            if (payloadBytes < 0)
+                            {
+                                throw new IOException(SR.net_frame_read_size);
+                            }
+                        }
 
-                    // Read in the rest of the payload if we don't have it.
-                    if (_internalBufferCount < payloadBytes)
-                    {
-                        await FillBufferAsync(adapter, payloadBytes).ConfigureAwait(false);
+                        // Read in the rest of the payload if we don't have it.
+                        if (_internalBufferCount < payloadBytes)
+                        {
+                            await FillBufferAsync(adapter, payloadBytes).ConfigureAwait(false);
+                        }
                     }
 
+                haveFullTlsFrame:
```

Comment:
It seems like we could use a loop here (probably do/while) rather than goto.

Reply:
I was thinking about it but at the end felt the goto is more readable. I can add the loop and intention if you feel that is better.

Reply:
In general it's pretty hard to follow the logic here. It's not just the goto, it's also the breaks and continues etc and trying to figure out where they are going to break/continue to.

Reply:
with the new structure the label is gone and we got back now via `continue`